pkg/container/qemu_runner.go

```diff
@@ -1188,3 +1115,187 @@ func randomPortN() (int, error) {
+	if _, err := os.Stat(initramfs); err == nil {
+		return initramfs, nil
+	}
+
+	clog.FromContext(ctx).Info("qemu: generating initramfs")
+
+	err := os.MkdirAll(filepath.Join(
+		"kernel",
+		apko_types.Architecture(runtime.GOARCH).ToAPK()),
+		os.ModePerm)
+	if err != nil {
+		return "", fmt.Errorf("unable to dest directory: %w", err)
+	}
+
+	spec := apko_types.ImageConfiguration{
+		Contents: apko_types.ImageContents{
+			RuntimeRepositories: []string{
+				"https://packages.wolfi.dev/os",
+			},
+			Keyring: []string{
+				"https://packages.wolfi.dev/os/wolfi-signing.rsa.pub",
+			},
+			Packages: []string{
+				"microvm-init",
+			},
+		},
+		Accounts: apko_types.ImageAccounts{
+			Groups: []apko_types.Group{{
+				GID:       1000,
+				GroupName: "build",
+			}},
+			Users: []apko_types.User{{
+				GID:      intToGID(1000),
+				HomeDir:  "/home/build",
+				UID:      1000,
+				UserName: "build",
+			}},
+		},
+	}
+	opts := []apko_build.Option{apko_build.WithImageConfiguration(spec),
+		apko_build.WithArch(apko_types.Architecture(runtime.GOARCH)),
+	}
+
+	tmpDir, err := os.MkdirTemp("", "melange-guest-*.initramfs")
+	if err != nil {
+		return "", fmt.Errorf("unable to create build context: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	bc, err := apko_build.New(ctx, apkofs.DirFS(tmpDir, apkofs.WithCreateDir()), opts...)
+	if err != nil {
+		return "", fmt.Errorf("unable to create build context: %w", err)
+	}
+
+	bc.Summarize(ctx)
+	if err := bc.BuildImage(ctx); err != nil {
+		return "", fmt.Errorf("unable to generate image: %w", err)
+	}
+	layerTarGZ, layer, err := bc.ImageLayoutToLayer(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(layerTarGZ)
+
+	log.Debugf("using %s for image layer", layerTarGZ)
+
+	// in case of some kernel images, we also need the /lib/modules directory to load
+	// necessary drivers, like 9p, virtio_net which are foundamental for the VM working.
+	if qemuModule, ok := os.LookupEnv("QEMU_KERNEL_MODULES"); ok {
+		clog.FromContext(ctx).Debugf("qemu: QEMU_KERNEL_MODULES env set, injecting modules in initramfs")
+		if _, err := os.Stat(qemuModule); err == nil {
+			clog.FromContext(ctx).Debugf("qemu: local QEMU_KERNEL_MODULES dir detected, injecting")
+			layer, err = injectKernelModules(ctx, layer, qemuModule)
+			if err != nil {
+				clog.FromContext(ctx).Errorf("qemu: could not inject needed kernel modules into initramfs: %v", err)
+				return "", err
+			}
+		}
+	}
+
+	guestInitramfs, err := os.Create(initramfs)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("failed to create cpio initramfs: %v", err)
+		return "", err
+	}
+
+	gzw := gzip.NewWriter(guestInitramfs)
+	if err := apko_cpio.FromLayer(layer, gzw); err != nil {
+		clog.FromContext(ctx).Errorf("failed to convert cpio initramfs: %v", err)
+		return "", err
+	}
+	if err := gzw.Close(); err != nil {
+		clog.FromContext(ctx).Errorf("failed to close gzip writer: %v", err)
+		return "", err
+	}
+
+	return guestInitramfs.Name(), nil
```

Comment:
I would skip the gzip since this initrd is always small and one of the key motivations for zipping before was to squeeze under the 2GB limit.

Reply:
Fixed